### ISSUE TITLE
Data race in gateway allocations

### ISSUE THREAD

**Issue by kalabukdima** [subsquid/archive.py#48]
With the raising number of gateways, a lot of workers started to crash with the error `sqlite3.IntegrityError: UNIQUE constraint failed: operators.address`.
<details><summary>Full log</summary>
<p>

```
Feb 27 09:50:22 WORKER subsquid-p2p-worker[2944284]: {"time": 1709027422479, "level": "ERROR", "ns": "asyncio", "msg": "Exception in callback <built-in method set_result of _asyncio.Future object at 0x7fa240239bd0>
handle: <Handle Future.set_result>", "err": "Traceback (most recent call last):
 File \"/usr/local/lib/python3.11/asyncio/runners.py\", line 190, in run
 return runner.run(main)
 ^^^^^^^^^^^^^^^^
 File \"/usr/local/lib/python3.11/asyncio/runners.py\", line 118, in run
 return self._loop.run_until_complete(task)
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File \"uvloop/loop.pyx\", line 1517, in uvloop.loop.Loop.run_until_complete
 File \"/app/env/lib/python3.11/site-packages/sqa/util/asyncio.py\", line 81, in run
 await program
 File \"/app/env/lib/python3.11/site-packages/sqa/worker/p2p/server.py\", line 348, in _main
 await monitor_service_tasks([
 File \"/app/env/lib/python3.11/site-packages/sqa/util/asyncio.py\", line 35, in monitor_service_tasks
 raise exception
 File \"/app/env/lib/python3.11/site-packages/sqa/worker/p2p/server.py\", line 86, in run
 await monitor_service_tasks([receive_task, send_logs_task], log=LOG)
 File \"/app/env/lib/python3.11/site-packages/sqa/util/asyncio.py\", line 35, in monitor_service_tasks
 raise exception
 File \"/app/env/lib/python3.11/site-packages/sqa/worker/p2p/server.py\", line 96, in _receive_loop
 await self._handle_query(peer_id, envelope.query, gateway_allocations)
 File \"/app/env/lib/python3.11/site-packages/sqa/worker/p2p/server.py\", line 165, in _handle_query
 if not await gateway_allocations.try_to_execute(client_peer_id):
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File \"/app/env/lib/python3.11/site-packages/sqa/gateway_allocations/gateway_allocations.py\", line 55, in try_to_execute
 self._storage.update_allocation(cluster, allocated_cus)
 File \"/app/env/lib/python3.11/site-packages/sqa/gateway_allocations/computation_units_storage.py\", line 108, in update_allocation
 self._db_conn.execute(
sqlite3.IntegrityError: UNIQUE constraint failed: operators.address
``` 

</p>
</details>

Indeed, we are checking for gateway allocations in parallel like this without any lock:
https://github.com/subsquid/archive.py/blob/a8b394283aab4616529ebd49ed9c70dc3f302359/sqa/gateway_allocations/gateway_allocations.py#L49-L55
Basically, three things happen here non-atomically:
- check if `gateway_id` exists in `gateways`, proceed if not
- insert cluster's operator info in `operators`
- for each gateway in the cluster insert or update it into `gateways`

Therefore, multiple routines can get to the second step and try to insert an entry with the same primary key.

Problems to fix:
- concurrent queries lead to a data race
- process crashes instead of failing a single query
